core/arch/arm/plat-rcar/romapi.c

```diff
@@ -0,0 +1,124 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2021, EPAM Systems
+ */
+#include <kernel/panic.h>
+#include <mm/core_memprot.h>
+#include <mm/core_mmu.h>
+
+#include "rcar.h"
+#include "romapi.h"
+
+static int get_api_table_index(void)
+{
+	/*
+	 * Depending on SoC type and version, there are 5 possible addresses
+	 * for each ROMAPI function
+	 */
+	static int index __nex_data = -1;
+
+	if (index != -1)
+		return index;
+
+	switch (rcar_get_product_type()) {
+	case RCAR_H3_ES1_0:
+	case RCAR_H3_ES1_1:
+		index = 0;
+		break;
+	case RCAR_H3_ES2_0:
+		index = 1;
+		break;
+	case RCAR_M3_ES1_00:
+	case RCAR_M3_ES1_05:
+		index = 2;
+		break;
+	case RCAR_H3_ES3_0:
+	case RCAR_M3_ES1_06:
+	case RCAR_M3_ES1_1:
+	case RCAR_V3M_ES_2_0:
+	case RCAR_D3:
+	case RCAR_E3:
+	case RCAR_M3N:
+	case RCAR_V3H:
+		index = 3;
+		break;
+	case RCAR_V3M_ES_1_0:
+		index = 4;
+		break;
+	default:
+		panic("Unknown RCAR product type");
+	}
+
+	return index;
+}
+
+/* implemented in romapi_call.S */
+extern uint32_t __plat_call_romapi(paddr_t sp, paddr_t func, uint64_t arg1,
+				   uint64_t arg2, uint64_t arg3);
+
+static uint32_t plat_call_romapi(paddr_t func, uint64_t arg1,
+				 uint64_t arg2, uint64_t arg3)
+{
+	/*
+	 * If MMU is enabled, we need to use trampoline function that will
```

Comment:
Wouldn't it be nice to flatmap the needed parts of the ROM?
That place to do that would probably be in `core_init_mmu_map()`. I'm sure we can do something that isn't too ugly.

Reply:
It was my first intention.
We need to add at least two new memory types: 
 - identity-mapped MMIO ranges (because ROM code accesses MMIO using absolute addresses)
 - identity-mapped non-TEE RX memory (for ROM code itself)

And this would require some rework of `core_mmu.c`. If you think that this changes are okay, then I'll rewrite this PR.

Reply:
With the MMIO ranges also, it might be a bit too much. My main concern is that we get too many ranges identity mapped. That could lead to `init_mem_map()` failing to find a suitable range if we're unlucky. Or leave us without a suitable location for the user mode table.

Another option could be to reimplement the hwrng driver in OP-TEE, but I guess you have your reasons to avoid that too. I'm fine with it as it is then.